sisyphus/sisyphusd_test.go

```diff
@@ -0,0 +1,216 @@
+		minRunNo := 999999
+		maxRunNo := 0
+		for k, v := range prowResultsForOneJobTmp {
+			i, err := strconv.Atoi(k)
+			if err != nil {
+				log.Fatalf("Error converting %s to int :%v", k, err)
+			}
+			if i > maxRunNo {
+				maxRunNo = i
+			}
+			if i < minRunNo {
+				minRunNo = i
+			}
+			prowResultsForOneJob[i] = v
+		}
+		mock.prowResults[job] = prowResultsForOneJob
+		mock.lastestRunNos[job] = minRunNo
+		mock.maxRunNos[job] = maxRunNo
+	}
+	return mock
+}
+
+func (p *ProwAccessorMock) GetLatestRun(jobName string) (int, error) {
+	ret := p.lastestRunNos[jobName]
+	p.lastestRunNos[jobName]++
+	if ret > p.maxRunNos[jobName] {
+		allRunsFinished := true
+		for j, lastestRun := range p.lastestRunNos {
+			if lastestRun <= p.maxRunNos[j] { // equal sign to ensure the last run is seen before termination
+				allRunsFinished = false
+			}
+		}
+		if allRunsFinished && p.exitSignalToSisyphus != nil && len(p.exitSignalToSisyphus) == 0 {
+			p.exitSignalToSisyphus <- true
+		}
+		return p.maxRunNos[jobName], nil
+	}
+	return ret, nil
+}
+
+func (p *ProwAccessorMock) GetProwResult(jobName string, runNo int) (*ProwResult, error) {
+	res := p.prowResults[jobName][runNo]
+	return &res, nil
+}
+
+func (p *ProwAccessorMock) GetGubernatorURL() string {
+	return p.gubernatorURL
+}
+
+func (p *ProwAccessorMock) Rerun(jobName string, runNo, numRerun int) error {
+	return nil
+}
+
+type StorageMock struct {
+	t             *testing.T
+	expectedStats []FlakeStat
+}
+
+func NewStorageMock(t *testing.T) *StorageMock {
+	dataFile := filepath.Join(testDataDir, expectedFlakeStatsJSON)
+	raw, err := ioutil.ReadFile(dataFile)
+	if err != nil {
+		log.Fatalf("Error reading %s:%v", dataFile, err)
+	}
+	var expectedStats []FlakeStat
+	if err = json.Unmarshal([]byte(raw), &expectedStats); err != nil {
+		log.Fatalf("Failed to unmarshal test data for %s: %v", dataFile, err)
+	}
+	return &StorageMock{
+		t:             t,
+		expectedStats: expectedStats,
+	}
+}
+
+func (s *StorageMock) Store(jobName, sha string, newFlakeStat FlakeStat) error {
+	for _, expectedStat := range s.expectedStats {
+		if expectedStat.TestName == jobName && expectedStat.SHA == sha {
+			if !reflect.DeepEqual(expectedStat, newFlakeStat) {
+				s.t.Errorf("Expecting %v but got %v", expectedStat, newFlakeStat)
+			}
+			return nil
+		}
+	}
+	s.t.Errorf("No matching expectedStat found given jobName = %s, sha = %s", jobName, sha)
+	return nil
+}
+
+func TestDaemonConfig(t *testing.T) {
+	catchFlakesByRun := true
+	cfg := &Config{
+		CatchFlakesByRun: catchFlakesByRun,
+	}
+	cfgExpected := &Config{
+		CatchFlakesByRun: catchFlakesByRun,
+		PollGapDuration:  DefaultPollGapDuration,
+		NumRerun:         DefaultNumRerun,
+	}
+	sisyphusd := NewDaemon(protectedJobsMock, prowProjectMock,
+		prowZoneMock, gubernatorURLMock, gcsBucketMock, cfg)
+	if !reflect.DeepEqual(sisyphusd.GetConfig(), cfgExpected) {
+		t.Error("setting catchFlakesByRun failed")
+	}
+}
+
+func TestProwResultsMock(t *testing.T) {
```

Comment:
funny that you have a test for you Mock

Reply:
I know *laugh_cry* The mock has become rather complected, and it becomes necessary to ensure the mock behaviors conform to its specification before we use the mock to test even more complicated rerun logic.